File: src/main/java/io/codeka/gaia/repository/StackRepository.java
package io.codeka.gaia.repository;

import io.codeka.gaia.bo.Stack;
import io.codeka.gaia.bo.StackState;
import io.codeka.gaia.teams.bo.Team;
import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

/**
 * Repository interface for stacks
 */
@Repository
public interface StackRepository extends MongoRepository<Stack, String>{

    int countStacksByStateAndOwnerTeam(StackState state, Team team);

    int countStacksByState(StackState state);

    List<Stack> findByOwnerTeam(Team team);

    Optional<Stack> findByIdAndOwnerTeam(String id, Team team);
}

File: src/main/java/io/codeka/gaia/stacks/repository/StackRepository.kt
package io.codeka.gaia.stacks.repository

import io.codeka.gaia.stacks.bo.Stack
import io.codeka.gaia.stacks.bo.StackState
import io.codeka.gaia.teams.Team
import io.codeka.gaia.teams.User
import org.springframework.data.mongodb.repository.MongoRepository
import org.springframework.stereotype.Repository
import java.util.*

/**
 * Repository interface for stacks
 */
@Repository
interface StackRepository : MongoRepository<Stack, String> {
    fun countStacksByStateAndOwnerTeam(state: StackState, team: Team): Long
    fun countStacksByStateAndCreatedBy(state: StackState, user: User): Long
    fun countStacksByState(state: StackState): Long
    fun findByCreatedBy(userWithNoTeam: User): List<Stack>
    fun findByOwnerTeam(team: Team): List<Stack>
    fun findByIdAndOwnerTeam(id: String, team: Team): Optional<Stack>
}
